File: plugins/modulation/vibrato/vibrato.js
// Vibrato — pure pitch wobble: variable delay only, no dry blend. By
// pulling samples from an LFO-swept tap, the playback rate of the
// buffered signal accelerates and decelerates, which the ear hears as
// frequency modulation. Unlike chorus (delay + dry), vibrato doesn't
// produce comb filtering — just the wobble itself.

function VibratoProcessor() {
  this.buf = new Array(4000);
  for (var k = 0; k < this.buf.length; k++) this.buf[k] = 0;
  this.bufLen = this.buf.length;
  this.writeIdx = 0;
  this.phase = 0;
}
VibratoProcessor.parameterDescriptors = [
  { name: 'rate',  label: 'Rate (Hz)',  defaultValue: 5, minValue: 0.5, maxValue: 12 },
  { name: 'depth', label: 'Depth (ms)', defaultValue: 2, minValue: 0,   maxValue: 8 }
];
VibratoProcessor.prototype.process = function(inputs, outputs, parameters) {
  var input = inputs[0][0], output = outputs[0][0];
  if (!input || !output) return true;
  var rate = parameters.rate[0];
  var depthMs = parameters.depth[0];
  var baseSamp = sampleRate * 0.005;
  var depthSamp = sampleRate * depthMs * 0.001;
  var phaseInc = (2 * Math.PI * rate) / sampleRate;
  for (var i = 0; i < input.length; i++) {
    var lfo = 0.5 + 0.5 * Math.sin(this.phase);
    var read = this.writeIdx - (baseSamp + depthSamp * lfo);
    while (read < 0) read += this.bufLen;
    var i0 = Math.floor(read);
    var frac = read - i0;
    var i1 = (i0 + 1) % this.bufLen;
    output[i] = this.buf[i0] * (1 - frac) + this.buf[i1] * frac;
    this.buf[this.writeIdx] = input[i];
    this.writeIdx = (this.writeIdx + 1) % this.bufLen;
    this.phase += phaseInc;
    if (this.phase > 2 * Math.PI) this.phase -= 2 * Math.PI;
  }
  return true;
};

registerProcessor('vibrato', VibratoProcessor);

File: plugins/modulation/vibrato/Vibrato.java
package com.vocalmonitor.plugin.community;

import com.vocalmonitor.plugin.VocalMonitorNativePlugin;

// Vibrato (native port) — pure pitch wobble: LFO-modulated delay tap,
// no dry blend. Unlike chorus this doesn't create comb-filter notches,
// just the seasick wobble itself.
public final class Vibrato implements VocalMonitorNativePlugin {
    private float[] buf;
    private int bufLen;
    private int writeIdx;
    private float phase;
    private int sampleRate = 44100;
    private float rate = 5f, depth = 2f;

    @Override
    public void init(int sr) {
        this.sampleRate = sr;
        bufLen = 4000;
        buf = new float[bufLen];
        writeIdx = 0;
        phase = 0f;
    }

    @Override public String[] parameterNames() { return new String[] { "rate", "depth" }; }
    @Override public float parameterMin(String n) { return "rate".equals(n) ? 0.5f : 0f; }
    @Override public float parameterMax(String n) { return "rate".equals(n) ? 12f : 8f; }
    @Override public float parameterDefault(String n) { return "rate".equals(n) ? 5f : 2f; }
    @Override public String parameterLabel(String n) { return "rate".equals(n) ? "Rate (Hz)" : "Depth (ms)"; }
    @Override public void setParameter(String n, float v) {
        if ("rate".equals(n)) rate = v;
        else if ("depth".equals(n)) depth = v;
    }

    @Override
    public void process(float[] input, float[] output) {
        final float baseSamp = sampleRate * 0.005f;
        final float depthSamp = sampleRate * depth * 0.001f;
        final float phaseInc = (float) (2.0 * Math.PI * rate / sampleRate);
        final float twoPi = (float) (2.0 * Math.PI);
        final float[] b = buf;
        final int bL = bufLen;
        int w = writeIdx;
        float ph = phase;
        final int n = input.length;
        for (int i = 0; i < n; i++) {
            float lfo = 0.5f + 0.5f * (float) Math.sin(ph);
            float read = w - (baseSamp + depthSamp * lfo);
            while (read < 0) read += bL;
            int i0 = (int) read;
            float frac = read - i0;
            int i1 = i0 + 1; if (i1 >= bL) i1 = 0;
            output[i] = b[i0] * (1f - frac) + b[i1] * frac;
            b[w] = input[i];
            w++; if (w >= bL) w = 0;
            ph += phaseInc;
            if (ph > twoPi) ph -= twoPi;
        }
        writeIdx = w;
        phase = ph;
    }
}
